apps/ejabberd/src/mod_aws_sns.erl

```diff
@@ -0,0 +1,293 @@
+%%%-------------------------------------------------------------------
+%%% @author Rafal Slota
+%%% @copyright (C) 2017 Erlang Solutions Ltd.
+%%% This software is released under the Apache License, Version 2.0
+%%% cited in 'LICENSE.txt'.
+%%% @end
+%%%-------------------------------------------------------------------
+%%% @doc
+%%% Amazon SNS notifications. This module gathers all message send by users and all presence
+%%% changes and publishes those events to AWS SNS.
+%%% @end
+%%%-------------------------------------------------------------------
+-module(mod_aws_sns).
+-author("Rafal Slota").
+
+-behavior(gen_mod).
+
+-include_lib("ejabberd/include/ejabberd.hrl").
+-include_lib("ejabberd/include/jlib.hrl").
+-include_lib("erlcloud/include/erlcloud_aws.hrl").
+
+-callback user_guid(UserJID :: ejabberd:jid()) -> mod_aws_sns:user_guid().
+-callback message_attributes(TopicARN :: mod_aws_sns:topic_arn(), UserJID :: ejabberd:jid(),
+                             IsOnline :: boolean()) -> mod_aws_sns:attributes().
+-callback message_attributes(TopicARN :: mod_aws_sns:topic_arn(), From :: ejabberd:jid(),
+                             To :: ejabberd:jid(), MessageType :: pm | muc,
+                             Packet :: jlib:xmlel()) -> mod_aws_sns:attributes().
+
+%%%===================================================================
+%%% Types and definitions
+%%%===================================================================
+
+-define(TOPIC_BASE, ["arn", "aws", "sns"]).
+
+-type user_guid() :: binary().
+-type topic_arn() :: string(). %% Full topic ARN in format arn:aws:sns:{REGION}:{ACCOUNT_ID}:{TOPIC}
+-type topic() :: string(). %% {TOPIC} part of topic_arn() type
+-type attributes() :: #{string() => string() | binary() | number()}.
+
+
+%%%===================================================================
+%%% Exports
+%%%===================================================================
+
+%% MIM module callbacks
+-export([start/2, stop/1]).
+
+%% Hooks
+-export([user_send_packet/3, user_present/1, user_not_present/4, filter_room_packet/2]).
+
+%% API
+-export([try_publish/5]).
+
+%% Types
+-export_type([user_guid/0, topic_arn/0, topic/0, attributes/0]).
+
+%%%===================================================================
+%%% API functions
+%%%===================================================================
+
+-spec start(Host :: ejabberd:server(), Opts :: proplists:proplist()) -> ok.
+start(Host, Opts) ->
+    MUCHost = gen_mod:get_opt_subhost(Host, muc_host, Opts, mod_muc:default_host()),
+    ejabberd_hooks:add(filter_room_packet, MUCHost, ?MODULE, filter_room_packet, 90),
```

Comment:
IMO, if it is a behaviour, then it should not attach handlers - I know it works because handlers are indexed by {Seq, Module, Function} tuple so there is no duplication, but still it doesn't seem to be the right place for it. Feel free to ask others, I'd also like to know what they think.